mongo-store.js

```diff
@@ -181,40 +185,163 @@ module.exports = function (opts) {
     },
 
     save: function (args, cb) {
-      var ent = args.ent
+      const is_update = Boolean(args.ent.id)
 
-      var update = !!ent.id
+      if (is_update) {
+        return updateExisting(args, cb)
+      }
 
-      getcoll(args, ent, function (err, coll) {
-        if (error(args, err, cb)) return
+      return createAndSave(args, cb)
 
-        var entp = ent.data$(false)
 
-        if (!update) {
-          var id
-          if (undefined !== ent.id$) {
-            id = ent.id$
-          } else if (opts.generate_id) {
-            id = opts.generate_id(ent)
+      function createAndSave(args, cb) {
+        return upsertIfRequested(args, function (err, upsert) {
+          if (error(args, err, cb)) {
+            return
           }
 
-          entp._id = makeid(id)
+          if (upsert.upsert_requested) {
+            return cb(null, upsert.out)
+          }
+
+          return createNew(args, cb)
+        })
+
+
+        function upsertIfRequested(args, cb) {
+          const query_for_save = args.q
+
+          if (Array.isArray(query_for_save.upsert$)) {
+            const upsert_on = query_for_save.upsert$
+            const public_entdata = args.ent.data$(false)
+
+            if (upsert_on.length > 0 && upsert_on.every(p => p in public_entdata)) {
+              const filter_by = upsert_on
+                .reduce((acc, field) => {
+                  acc[field] = args.ent[field]
+                  return acc
+                }, {})
+
+              const replacement = (() => {
+                const o = Dot.flatten(Object.assign({}, public_entdata))
+
+
+                const NO_SPECIFIC_ID_REQUESTED = {}
```

Comment:
why does this replace:
```
          if (undefined !== ent.id$) {
            id = ent.id$
          } else if (opts.generate_id) {
            id = opts.generate_id(ent)
          }

          entp._id = makeid(id)
```

Reply:
@rjrodger I believe the approach in this PR to be more readable as it hides the details of how it tries to retrieve the id in a small tidy function.

The problem I found with the original code was that it was not clear off the bat that it was intentional that `id` may be undefined. For a moment I actually thought there might have been a bug.

Besides, I strongly believe that `undefined` as a value is not declarative in code. I would personally refrain from using `undefined` in code, where appropriate. The alternative present in this PR is readable and communicates the intention to the reader arguably more clearly.

If you would like to inline the original code back, may I recommend a consensus:
```javascript

let base_id

if (null === ent.id$ || undefined === ent.id$) {
  base_id = ent.id$
} else if (opts.generate_id) {
  base_id = opts.generate_id(ent)
} else {
  base_id = undefined
}

const new_id = makeid(base_id)
entp._id = new_id
```